Add Nobø Hub week profiles and global override

oyvindwe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Implemented `select` entities for:
- Selecting week profile for a zone.
- Overriding preset for all zones (global override).

Selecting week profile is useful for a workaround  to turn all heaters in a zone to off, which can only be done via a week profile (auto mode). It may also be useful to have automations that switch week profiles based on events, e.g. have a different week profile for staying at home on a weekday, but still lower temperature at night.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24675

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

oyvindwe committed:
* Nobø Ecohub select entities
  - Week profiles
  - Global overrides

oyvindwe committed:
Set integration_type

oyvindwe commented on homeassistant/components/nobo_hub/select.py:
```diff
@@ -0,0 +1,168 @@
+        else nobo.API.OVERRIDE_TYPE_CONSTANT
+    )
+
+    entities: list[SelectEntity] = [
```
The list will in any case be short - typically at most one zone per room of a house.

oyvindwe committed:
Typing

oyvindwe committed:
Remove translations

cbjornerud commented:
Installed and tested on HA 2023.3.5 running om HA Yellow. This is great! I can see active week profile for all my Nobo devices - and I can change active week profile with an automation. I have an automation to save power on heating when a window is open. Previously this would select pre-set Eco - which sets the thermostat to 7 degrees C. Now the automation can choose any week profile. As Nobo supports empty week profiles, this allows me to turn off the heat completely - which is what I want as long as the window is open.
This should be a standard part of the Nobo integration.

oyvindwe committed:
Translation fixes
- Moved strings.select.json into strings.json
- Added translation keys for select entities
- Use shared translation keys for global override states

oyvindwe committed:
Use DeviceInfo object

oyvindwe committed:
Revert temperature name - uses device class name

oyvindwe committed:
Fix updated checks

oyvindwe committed:
Improve error handling (preparing for Silver level)

emontnemery commented on homeassistant/components/nobo_hub/select.py:
```diff
@@ -0,0 +1,169 @@
+        self._attr_unique_id = f"{hub.hub_serial}"
+        self._override_type = override_type
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, f"{hub.hub_serial}")},
```
same

oyvindwe replied:
Done

emontnemery commented on homeassistant/components/nobo_hub/select.py:
```diff
@@ -0,0 +1,169 @@
+        self._nobo = hub
+        self._attr_unique_id = f"{hub.hub_serial}:{zone_id}:profile"
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, f"{hub.hub_serial}:{zone_id}")},
```
Do we really need a separate device for each zone, shouldn't these be included in the hub device?

oyvindwe replied:
This is how I was adviced to do it for the climate entity. Each zone can consist of one or more heaters or floor thermostats, so the actual physical devices are not included. This identifier matches the device info in the climate entity.

(I've since discovered that some floor thermostats requires a separate zone. A possibility is to use the actual device in device info when a zone only consists of one device, but this would require a similar change in the climate entity. We would also need to decide what device (or none) to use for zones that consists of multiple devices.)

emontnemery replied:
Ah, I didn't realize this device id is already used in the climate entity, in that case it's fine 👍

emontnemery commented on homeassistant/components/nobo_hub/select.py:
```diff
@@ -0,0 +1,169 @@
+    def _read_state(self) -> None:
+        for override in self._nobo.overrides.values():
+            if override["target_type"] == nobo.API.OVERRIDE_TARGET_GLOBAL:
+                self._attr_current_option = self._modes[override["mode"]]
```
shouldn't we break out of the for loop?
```suggestion
                self._attr_current_option = self._modes[override["mode"]]
                break
```

oyvindwe replied:
Done

emontnemery reviewed (commented):
Looks good, just a comment on the device structure

oyvindwe commented on homeassistant/components/nobo_hub/select.py:
```diff
@@ -0,0 +1,169 @@
+    def __init__(self, hub: nobo, override_type) -> None:
+        """Initialize the global override selector."""
+        self._nobo = hub
+        self._attr_unique_id = f"{hub.hub_serial}"
```
Done

oyvindwe committed:
Review

emontnemery reviewed (approved):
Thanks, @oyvindwe 👍
